### PR TITLE
[TR] PHP7 and HHVM compatibility

### PR DESCRIPTION
**Pull request by nidup**
I continued the PR opened here https://github.com/akeneo/pim-community-dev/pull/3525 related to #3343 

With the fix written by @jjanvier (related to reserved keywords) and by bumping Phpspec to 2.4, we achieve to run with success our unit tests on php7 and hhvm engines.

unit-tests-matrix

Due to weird behavior between phpspec and php7 when defining concret classes in a phpspec file to test an abstract class (method signatures detected as not respecting interfaces methods), i dropped 3 phpspec to integrate them to related base class spec (means from AbstractAttributeCopierSpec to BaseAttributeCopierSpec for instance).

The PR also remove PHP7 and HHVM from the allowed failures matrix of Travis CI.

You can notice that we still have an issue with PHP driver for MongoDB for both PHP7 and HHVM.

| Q                 | A
| ----------------- | ---
| Specs             | N
| Behats            | N
| Blue CI           | Y
| Changelog updated | Y
| Review and 2 GTM  | ?

### CODE REVIEW

**Comment by nidup** [comment]
Still an issue with ```PHP Fatal error:  Uncaught TypeError: Argument 1 passed to spec\Akeneo\Bundle\BatchBundle\Connector\ConnectorRegistrySpec::let() must be an instance of Akeneo\Bundle\BatchBundle\Job\JobFactory, instance of PhpSpec\Wrapper\Collaborator given in /home/travis/build/akeneo/pim-community-dev/spec/Akeneo/Bundle/BatchBundle/Connector/ConnectorRegistrySpec.php:14``` looks related to https://github.com/phpspec/phpspec/issues/659

If related to this, issue has been fixed in phpspec 2.4-alpha

File: .travis.yml [hunk 1]
```diff
@@ -27,15 +22,15 @@ before_script:
     - sh -c 'if [ "$TRAVIS_PHP_VERSION" != "hhvm" ]; then composer update --prefer-dist --no-interaction; else composer update --prefer-dist --no-interaction --no-scripts; fi;'
     - sh -c 'if [ "$TRAVIS_PHP_VERSION" != "hhvm" ]; then ./app/console oro:requirejs:generate-config; fi;'
     - sh -c 'if [ "$TRAVIS_PHP_VERSION" != "hhvm" ]; then ./app/console assets:install; fi;'
-    - npm install -g grunt-cli
-    - npm install
-    - curl http://get.sensiolabs.org/php-cs-fixer.phar -o php-cs-fixer
+    - sh -c 'if [ "$TRAVIS_PHP_VERSION" = "5.4" ]; then npm install -g grunt-cli; fi;'
+    - sh -c 'if [ "$TRAVIS_PHP_VERSION" = "5.4" ]; then npm install; fi;'
+    - sh -c 'if [ "$TRAVIS_PHP_VERSION" = "5.4" ]; then curl http://get.sensiolabs.org/php-cs-fixer.phar -o php-cs-fixer; fi;'
```
Comment (ClementGautier) [inline .travis.yml]: Why do we run grunt and phpcs on php 5.4 only ? Tomorrow if we decide to not support 5.4 anymore we will have to think about changing this line. I would rather launch then in every environments to ensure there is no clue with the tools in this versions.

File: spec/Pim/Component/Localization/Localizer/LocalizedAttributeConverterSpec.php [hunk 2]
```diff
@@ -67,6 +68,10 @@ function it_returns_a_constraint_validation_if_decimal_separator_is_not_expected
 
         $this->convertLocalizedToDefaultValues(['number' => [['data' => '10,45']]], $options)
             ->shouldReturn(['number' => [['data' => '10.45']]]);
-        $this->getViolations()->shouldNotBeNull();
+
+        $violations = $this->getViolations()->getWrappedObject();
+        if (null === $violations) {
```
Comment (ClementGautier) [inline spec/Pim/Component/Localization/Localizer/LocalizedAttributeConverterSpec.php]: I don't understand this change, you should probably add a comment here if it's the only way to deal with that but it really looks ugly and seems to do the same thing than `$this->getViolations()->shouldNotBeNull()`. Whats the differnce ?

**Comment by ClementGautier** [comment]
And please, squash your commits ;)

File: spec/Pim/Component/Localization/Localizer/LocalizedAttributeConverterSpec.php (hunk 2 above)
Comment (nidup) [inline spec/Pim/Component/Localization/Localizer/LocalizedAttributeConverterSpec.php]: @jjanvier you remember why this change ?

File: .travis.yml (hunk 1 above)
Comment (nidup) [inline .travis.yml]: Ok we need to think about maintainability vs perf improv.
It looks weird for me to run js sanitization for each php version.

Comment (grena) [inline .travis.yml]: Why do we install npm dependencies on travis? I don't think we use JS related stuff (`npm install -g grunt-cli` & `npm install`) I don't know if it's used by travis (specs & static analysis)

**Comment by grena** [comment]
Comment/Question, then GTM

**Comment by nidup** [comment]
Following @ClementGautier's feedback about maintainability issues, I reverted the change of Travis conf related to php 5.4, even better! I simplified it to remove useless hhvm exclusions/weirdness (now, we only exclude mongo setup for hhvm and php7 because it's not supported).

**Comment by nidup** [comment]
Comments have been fixed up, commits are squashed

**Comment by jjanvier** [comment]
Gtm
